### ISSUE TITLE
JSpecifyExperimental: Support smart cast for Class.getAnnotation after Class.isAnnotationPresent check

### ISSUE THREAD

**Issue by toellrich** [uber/NullAway#1766]
### Description
With the introduction of the new `-XepOpt:NullAway:JSpecifyExperimental=true` flag in version `0.14.0`, `Class.getAnnotation` is correctly treated as returning a `@Nullable` type. 

However, NullAway currently does not recognize when a `Class.getAnnotation(...)` call is guarded by a preceding `Class.isAnnotationPresent(...)` check for the exact same annotation class. This results in a false positive unless wrapped in `Objects.requireNonNull` (or similar).

As discussed with the maintainer on Discord, this issue tracks adding support for this specific conditional check pattern (ensuring the arguments to both calls match).

### Code Example

```java
import java.util.Objects;

@Deprecated(forRemoval = true)
public class Tester {

  public static void main(String[] args) {
    if (Tester.class.isAnnotationPresent(Deprecated.class)) {
      // NullAway reports a warning here if Objects.requireNonNull is omitted,
      // even though the presence of the annotation was already verified.
      boolean forRemoval = Tester.class.getAnnotation(Deprecated.class).forRemoval();
      System.out.println("forRemoval=" + forRemoval);
    }
  }
}
```

### Expected Behavior
When `-XepOpt:NullAway:JSpecifyExperimental=true` is active, NullAway should infer that `Tester.class.getAnnotation(Deprecated.class)` returns a `@NonNull` value inside the `if` block, because `Tester.class.isAnnotationPresent(Deprecated.class)` evaluated to `true`.

### Context
* **NullAway Version:** 0.14.0+
* **Flag:** `-XepOpt:NullAway:JSpecifyExperimental=true`

**Comment by msridhar** [comment]
Thanks for the report, @toellrich!  I think this will be a bit challenging to support safely in NullAway, due to the fact that the parameter and receiver arguments need to match.  I'd like to think more about it, but to be open we have higher priority issues to fix.

But, I kind of feel that in the presence of a nullability type system, methods like `isAnnotationPresent` may be an anti-pattern.  Are you open to rewriting your code like this?
```java
    var annot = Tester.class.getAnnotation(Deprecated.class);
    if (annot != null) {
      boolean forRemoval = annot.forRemoval();
      System.out.println("forRemoval=" + forRemoval);
    }
```
With the nullness modeling, you can be sure that NullAway will check you don't dereference the result of `getAnnotation` without checking its nullness.  So arguably this might be cleaner?  It doesn't work, though, if `getAnnotation` can return something non-null even when `isAnnotationPresent` would return false.

**Comment by toellrich** [comment]
Thanks for the reply, @msridhar! Wouldn't the support of `isAnnotationPresent` be similar to the following bit of code, though, which NullAway does support?

```java
  public static void main(String[] args) {
    Map<String, String> map = Map.of("1", "one", "2", "two");
    if (map.containsKey("1")) {
      System.out.println(map.get("1").length());
    }
  }
```

**Comment by msridhar** [comment]
Hi @toellrich the map case is a bit different, for two resons:

1. `Map.get` can return null for a key in the map, even in the case where `containsKey` returns true.  We don't support reporting warnings properly for that case yet, but we should; I opened #1805 on that.
2. Maps are used very commonly, justifying special handling.

To be open, we probably won't get to fixing this one soon, but I'll keep it open in case we're convinced otherwise.
